File: Week3/Code/get_TreeHeight.py
"""Python version of get_TreeHeight.R"""

__author__ = 'Matthew Campos (matthew.campos19@imperial.ac.uk)'
__version__ = '0.0.1'

import sys
import pandas as pd
import numpy as np
import re

def open_file(path):
    """Takes input file and outputs the data using pandas"""
    data = pd.read_csv(path) #reads the csv file
    return(data)

def Tree_height(data):
    """calculates tree height of the input file"""
    degrees = data['Angle.degrees']
    distance = data['Distance.m']
    radians = np.deg2rad(degrees) #convert degrees to radians
    height = distance * np.tan(radians)
    data['height'] = height
    return data

def writefile(data, path):
    """formats the file name of the result file"""
    file = re.findall(r"/+([\w\d]+).csv", sys.argv[1]) #finds csv input files
    filename = "Py_" + file[0] + r"_treeheights.csv"
    path = "../Results/" + filename
    data.to_csv(path, index=False) #function that writes it as a csv

def main(argv):
    """takes input file and runs the script"""
    data = open_file(sys.argv[1])
    output = Tree_height(data)
    writefile(output, sys.argv[1])

if (__name__=="__main__"):
    status = main(sys.argv)
    sys.exit("Exit")

File: Week3/Code/get_TreeHeight.R
# This function calculates heights of trees given distance of each tree 
# from its base and angle to its top, using  the trigonometric formula 
#
# height = distance * tan(radians)
#
# ARGUMENTS
# degrees:   The angle of elevation of tree
# distance:  The distance from base of tree (e.g., meters)
#
# INPUT
# input file entered in command line
# OUTPUT
# The heights of the tree, same units as "distance"

#!/usr/bin/env Rscript

##__author__ = 'Matthew Campos (matthew.campos19@imperial.ac.uk)'
##__version__ = '0.0.1'

args = commandArgs(trailingOnly=TRUE)

#checks if there is an input in the command line
if (length(args)==0){
  stop("At least one argument must be supplied (input file).n. Please check Data directory folder")
}

#sets the file to a variable
MyData <- read.csv(args[1], header=TRUE)
print(MyData)

#figure out how many distance measurements were taken and transfer to a vector
distance_vector <- vector()
for (i in MyData[2]){
  distance_vector <- c(i)
}
print(length(distance_vector)) 

#count angle degrees measurements and put them in a vector
angle_degrees_vector <- vector()
for (y in MyData[3]){
  angle_degrees_vector <- c(y)
}
print(length(angle_degrees_vector)) 

#calculates the tree height
TreeHeight <- function(degrees, distance){ 
  #calculates height using trignometric rules
  radians <- degrees * pi / 180
  height <- distance * tan(radians)
  #print(paste("Tree height is:", height))
  
  return (height)
}

#runs function to calculate tree height and assigns values into vector
Tree_height.m <- vector() #creates empty vector
for (z in 1:120){
  Tree_height.m <- c(Tree_height.m,TreeHeight(angle_degrees_vector[z],distance_vector[z]))
}

#Edit output name
file <- args[1] 
output_name <- sub(pattern = "(.*)\\..*$", replacement = "\\1", basename(file)) #removes file path and extension
print(output_name)

#creating file output
path_out = "../Results/" #create relative path
filename <- paste(path_out,output_name,"_treeheight.csv ", sep = "")

MyData <- cbind(MyData,Tree_height.m) #adds column of the vector of tree height to 
write.csv(MyData, filename)
